File: python/2022-11-13/leetcode124.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    res = float('-inf')

    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        def dfs(node: Optional[TreeNode]):
            if node is None:
                return 0
            left = dfs(node.left)
            right = dfs(node.right)
            one_side = max(node.val + left, node.val + right)
            two_side = node.val + left + right
            no_side = node.val
            self.res = max(self.res, one_side, two_side, no_side)

            return max(one_side, no_side)

        dfs(root)
        return self.res

File: cpp/2023-1-3/leetcode124.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
};

bool dfs(struct TreeNode* left, struct TreeNode* right) {
    if (left != NULL && right != NULL) {
        if (left->val == right->val)
            return dfs(left->left, right->right) &&
                   dfs(left->right, right->left);
    } else if (left == NULL && right == NULL)
        return true;
    return false;
}
bool isSymmetric(struct TreeNode* root) {
    if (root == NULL) return false;
    return dfs(root->left, root->right);
}
int main() {
    printf("Hello world\n");
    return 0;
}

